File: Python/hello_world/sh1107.py
# SH1107 OLED Display Driver for MicroPython
# For 128x128 monochrome OLED displays like Adafruit 1.12" OLED

import framebuf
from machine import I2C

# SH1107 commands
SET_CONTRAST = 0x81
SET_NORM_INV = 0xA6
SET_DISP = 0xAE
SET_DISP_START_LINE = 0xDC
SET_PAGE_ADDR = 0xB0
SET_COL_ADDR_LOW = 0x00
SET_COL_ADDR_HIGH = 0x10
SET_SEG_REMAP = 0xA0
SET_MUX_RATIO = 0xA8
SET_COM_OUT_DIR = 0xC0
SET_DISP_OFFSET = 0xD3
SET_DISP_CLK_DIV = 0xD5
SET_PRECHARGE = 0xD9
SET_VCOM_DESEL = 0xDB
SET_ENTIRE_ON = 0xA4


class SH1107_I2C(framebuf.FrameBuffer):
    def __init__(self, width: int, height: int, i2c: I2C, addr: int = 0x3D, rotate: int = 90):
        self.i2c = i2c
        self.addr = addr
        self.width = width
        self.height = height
        self.rotate = rotate

        # Buffer for display data
        self.buffer = bytearray(width * height // 8)

        # For 90/270 rotation, swap width/height for framebuffer
        if rotate in (90, 270):
            super().__init__(  # pyright: ignore[reportUnknownMemberType]
                self.buffer, height, width, framebuf.MONO_VLSB
            )
        else:
            super().__init__(  # pyright: ignore[reportUnknownMemberType]
                self.buffer, width, height, framebuf.MONO_VLSB
            )

        self._init_display()

    def _write_cmd(self, cmd: int) -> None:
        self.i2c.writeto(  # pyright: ignore[reportUnknownMemberType]
            self.addr, bytes([0x00, cmd])
        )

    def _write_data(self, data: bytes | bytearray) -> None:
        self.i2c.writeto(  # pyright: ignore[reportUnknownMemberType]
            self.addr, bytes([0x40]) + data
        )

    def _init_display(self) -> None:
        # Turn display off
        self._write_cmd(SET_DISP | 0x00)

        # Set clock divider
        self._write_cmd(SET_DISP_CLK_DIV)
        self._write_cmd(0x51)

        # Set multiplex ratio
        self._write_cmd(SET_MUX_RATIO)
        self._write_cmd(0x7F)  # 128 MUX

        # Set display offset
        self._write_cmd(SET_DISP_OFFSET)
        self._write_cmd(0x00)

        # Set display start line
        self._write_cmd(SET_DISP_START_LINE)
        self._write_cmd(0x00)

        # Set segment remap based on rotation
        if self.rotate in (0, 90):
            self._write_cmd(SET_SEG_REMAP | 0x00)
        else:
            self._write_cmd(SET_SEG_REMAP | 0x01)

        # Set COM output scan direction based on rotation
        if self.rotate in (0, 270):
            self._write_cmd(SET_COM_OUT_DIR | 0x08)
        else:
            self._write_cmd(SET_COM_OUT_DIR | 0x00)

        # Set contrast
        self._write_cmd(SET_CONTRAST)
        self._write_cmd(0x80)

        # Set precharge period
        self._write_cmd(SET_PRECHARGE)
        self._write_cmd(0x22)

        # Set VCOM deselect level
        self._write_cmd(SET_VCOM_DESEL)
        self._write_cmd(0x35)

        # Entire display on (use RAM)
        self._write_cmd(SET_ENTIRE_ON)

        # Normal display (not inverted)
        self._write_cmd(SET_NORM_INV)

        # Turn display on
        self._write_cmd(SET_DISP | 0x01)

        print("SH1107 display initialized")

    def show(self) -> None:
        """Update the display with the buffer contents"""
        for page in range(self.height // 8):
            self._write_cmd(SET_PAGE_ADDR | page)
            self._write_cmd(SET_COL_ADDR_LOW | 0)
            self._write_cmd(SET_COL_ADDR_HIGH | 0)

            start = page * self.width
            end = start + self.width
            self._write_data(self.buffer[start:end])

    def poweroff(self) -> None:
        self._write_cmd(SET_DISP | 0x00)

    def poweron(self) -> None:
        self._write_cmd(SET_DISP | 0x01)

    def contrast(self, value: int) -> None:
        self._write_cmd(SET_CONTRAST)
        self._write_cmd(value & 0xFF)

    def invert(self, invert: bool) -> None:
        self._write_cmd(SET_NORM_INV | (1 if invert else 0))

File: Rust/fitness-tracker/src/tasks/sh1107.rs
use embedded_graphics::draw_target::DrawTarget;
use embedded_graphics::geometry::{OriginDimensions, Size};
use embedded_graphics::pixelcolor::BinaryColor;
use embedded_graphics::Pixel;
use embedded_hal::i2c::I2c;

const ADDR: u8 = 0x3D;
const WIDTH: u32 = 128;
const HEIGHT: u32 = 128;
const PAGES: usize = (HEIGHT as usize) / 8;
const BUF_SIZE: usize = (WIDTH as usize) * (HEIGHT as usize) / 8;

const CMD_PREFIX: u8 = 0x00;
const DATA_PREFIX: u8 = 0x40;

pub struct Sh1107<I> {
    i2c: I,
    buf: [u8; BUF_SIZE],
}

impl<I: I2c> Sh1107<I> {
    pub fn new(i2c: I) -> Self {
        Self {
            i2c,
            buf: [0u8; BUF_SIZE],
        }
    }

    pub fn init(&mut self) -> Result<(), I::Error> {
        // Adafruit SH1107 128x128 init sequence
        let cmds: &[u8] = &[
            0xAE, // display off
            0xD5, 0x51, // set display clock div
            0x20, // set memory mode (page addressing)
            0x81, 0x4F, // set contrast
            0xAD, 0x8A, // internal IREF
            0xA0, // segment remap = 0
            0xC0, // COM scan normal
            0xDC, 0x00, // display start line = 0
            0xD3, 0x00, // display offset = 0
            0xD9, 0x22, // pre-charge period
            0xDB, 0x35, // VCOM deselect level
            0xA8, 0x7F, // multiplex ratio = 127
            0xA4, // entire display on (follow RAM)
            0xA6, // normal display (not inverted)
            0xAF, // display on
        ];
        self.send_commands(cmds)
    }

    pub fn clear_buffer(&mut self) {
        self.buf.fill(0);
    }

    pub fn flush(&mut self) -> Result<(), I::Error> {
        for page in 0..PAGES {
            self.send_commands(&[
                0xB0 | (page as u8), // set page address
                0x00,                // lower column address = 0
                0x10,                // upper column address = 0
            ])?;

            let start = page * (WIDTH as usize);
            let end = start + (WIDTH as usize);

            // Send data in a single I2C write: prefix byte + 128 bytes of page data
            let mut data_buf = [0u8; 1 + WIDTH as usize];
            data_buf[0] = DATA_PREFIX;
            data_buf[1..].copy_from_slice(&self.buf[start..end]);
            self.i2c.write(ADDR, &data_buf)?;
        }
        Ok(())
    }

    fn send_commands(&mut self, cmds: &[u8]) -> Result<(), I::Error> {
        for &cmd in cmds {
            self.i2c.write(ADDR, &[CMD_PREFIX, cmd])?;
        }
        Ok(())
    }
}

impl<I: I2c> OriginDimensions for Sh1107<I> {
    fn size(&self) -> Size {
        Size::new(WIDTH, HEIGHT)
    }
}

impl<I: I2c> DrawTarget for Sh1107<I> {
    type Color = BinaryColor;
    type Error = core::convert::Infallible;

    fn draw_iter<P>(&mut self, pixels: P) -> Result<(), Self::Error>
    where
        P: IntoIterator<Item = Pixel<Self::Color>>,
    {
        for Pixel(point, color) in pixels {
            let x = point.x;
            let y = point.y;
            if x < 0 || x >= WIDTH as i32 || y < 0 || y >= HEIGHT as i32 {
                continue;
            }
            let x = x as usize;
            let y = y as usize;
            let page = y / 8;
            let bit = y % 8;
            let idx = page * (WIDTH as usize) + x;
            match color {
                BinaryColor::On => self.buf[idx] |= 1 << bit,
                BinaryColor::Off => self.buf[idx] &= !(1 << bit),
            }
        }
        Ok(())
    }
}
